src/v/cluster/topic_table.cc

```diff
@@ -106,49 +106,105 @@ ss::future<> topic_table::stop() {
               "non_replicable_topic should exist in hierarchy: {}",
               tp_nsv);
         } else {
             /// Prevent deletion of source topics that have non_replicable
             /// topics. To delete this topic all of its non_replicable descedent
             /// topics must first be deleted
-            auto found = _topics_hierarchy.find(cmd.value);
+            auto found = _topics_hierarchy.find(nt);
             if (found != _topics_hierarchy.end() && !found->second.empty()) {
-                return ss::make_ready_future<std::error_code>(
-                  errc::source_topic_still_in_use);
+                return std::error_code(errc::source_topic_still_in_use);
             }
         }
 
         for (auto& p_as : tp->second.get_assignments()) {
             _partition_count--;
-            auto ntp = model::ntp(cmd.key.ns, cmd.key.tp, p_as.id);
+            auto ntp = model::ntp(nt.ns, nt.tp, p_as.id);
             _updates_in_progress.erase(ntp);
             _pending_deltas.emplace_back(
               std::move(ntp), p_as, offset, delete_type);
         }
 
         _topics.erase(tp);
         notify_waiters();
-        _probe.handle_topic_deletion(cmd.key);
+        _probe.handle_topic_deletion(nt);
 
-        return ss::make_ready_future<std::error_code>(errc::success);
+        return errc::success;
+    }
+
+    return errc::topic_not_exists;
+}
+
+ss::future<std::error_code>
+topic_table::apply(topic_lifecycle_transition soft_del, model::offset offset) {
+    _last_applied_revision_id = model::revision_id(offset);
+
+    if (soft_del.mode == topic_lifecycle_transition_mode::pending_gc) {
+        // Create tombstone
+        auto tp = _topics.find(soft_del.topic.nt);
+        if (tp == _topics.end()) {
+            return ss::make_ready_future<std::error_code>(
+              errc::topic_not_exists);
+        }
+
+        auto tombstone = nt_lifecycle_marker{
+          .config = tp->second.get_configuration(),
+          .initial_revision_id = tp->second.get_remote_revision().value_or(
+            model::initial_revision_id(tp->second.get_revision()))};
+
+        _lifecycle_markers.emplace(soft_del.topic, tombstone);
+        vlog(
+          clusterlog.debug,
+          "Created tombstone for topic {} {}",
+          soft_del.topic.nt,
+          soft_del.topic.initial_revision_id);
+    } else if (soft_del.mode == topic_lifecycle_transition_mode::drop) {
+        if (_lifecycle_markers.contains(soft_del.topic)) {
+            vlog(
+              clusterlog.debug,
+              "Purged tombstone for {} {}",
+              soft_del.topic.nt,
+              soft_del.topic.initial_revision_id);
+            _lifecycle_markers.erase(soft_del.topic);
+            return ss::make_ready_future<std::error_code>(errc::success);
+        } else {
+            // TOOD: should we assert here?  Or ignore the error?  In principle
+            // it is harmless, but indicates a bug.
```

Comment:
I think returning an error is fine. I'd also log at ERROR level here.

Reply:
Added error log